sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/interpreter/operator/StringOperators.java

```diff
@@ -0,0 +1,242 @@
+              SqlTypeName.VARCHAR, SqlFunctions.concat(left.getString(), right.getString()));
+        }
+      };
+
+  public static final BeamSqlOperator POSITION =
+      new StringOperator() {
+        @Override
+        public boolean accept(List<BeamSqlExpression> operands) {
+          if (operands.size() < 2 || operands.size() > 3) {
+            return false;
+          }
+
+          return SqlTypeName.CHAR_TYPES.contains(operands.get(0).getOutputType())
+              && SqlTypeName.CHAR_TYPES.contains(operands.get(1).getOutputType())
+              && ((operands.size() < 3)
+                  || SqlTypeName.INT_TYPES.contains(operands.get(2).getOutputType()));
+        }
+
+        @Override
+        public BeamSqlPrimitive apply(List<BeamSqlPrimitive> arguments) {
+          String targetStr = arguments.get(0).getString();
+          String containingStr = arguments.get(1).getString();
+          int from = arguments.size() < 3 ? 1 : arguments.get(2).getInteger();
+          return BeamSqlPrimitive.of(
+              SqlTypeName.INTEGER, SqlFunctions.position(targetStr, containingStr, from));
+        }
+      };
+
+  public static final BeamSqlOperator TRIM =
+      new StringOperator() {
+        @Override
+        public boolean accept(List<BeamSqlExpression> subexpressions) {
+          return (subexpressions.size() == 1
+                  && SqlTypeName.CHAR_TYPES.contains(subexpressions.get(0).getOutputType()))
+              || (subexpressions.size() == 3
+                  && SqlTypeName.SYMBOL.equals(subexpressions.get(0).getOutputType())
+                  && SqlTypeName.CHAR_TYPES.contains(subexpressions.get(1).getOutputType())
+                  && SqlTypeName.CHAR_TYPES.contains(subexpressions.get(2).getOutputType()));
+        }
+
+        @Override
+        public BeamSqlPrimitive apply(List<BeamSqlPrimitive> operands) {
+          SqlTrimFunction.Flag type;
+          String removeChars;
+          String fromStr;
+
+          if (operands.size() == 1) {
+            type = SqlTrimFunction.Flag.BOTH;
+            removeChars = " ";
+            fromStr = operands.get(0).getString();
+          } else {
+            type = (SqlTrimFunction.Flag) operands.get(0).getValue();
+            removeChars = operands.get(1).getString();
+            fromStr = operands.get(2).getString();
+          }
+
+          return BeamSqlPrimitive.of(
+              SqlTypeName.VARCHAR,
+              trim(
+                  type == SqlTrimFunction.Flag.BOTH || type == SqlTrimFunction.Flag.LEADING,
+                  type == SqlTrimFunction.Flag.BOTH || type == SqlTrimFunction.Flag.TRAILING,
+                  removeChars,
+                  fromStr));
+        }
+      };
+
+  /**
+   * Calcite's implementation of TRIM is incorrect and only trims the first character in removeStr.
+   *
+   * <p>This implementation deliberately kept compatible with eventual upstreaming.
+   */
+  private static String trim(
```

Comment:
isn't there a Guava trim method?

Reply:
I could use it. This is a copy/paste of Calcite's and I will contribute upstream. Then we are sure to have the "SQL" definition of trim. Many of the operators here were incorrect because they did not do things the SQL way.